src/property.hpp

```diff
@@ -19,145 +19,218 @@
 #ifndef REALM_PROPERTY_HPP
 #define REALM_PROPERTY_HPP
 
-#include <string>
-
 #include "util/compiler.hpp"
+#include "util/tagged_bool.hpp"
+
+#include <string>
 
 namespace realm {
-    enum class PropertyType : unsigned char {
-        Int    = 0,
-        Bool   = 1,
-        Float  = 9,
-        Double = 10,
-        String = 2,
-        Data   = 4,
-        Any    = 6, // deprecated and will be removed in the future
-        Date   = 8,
-        Object = 12,
-        Array  = 13,
-        LinkingObjects  = 14,
-    };
-
-    static const char *string_for_property_type(PropertyType type);
-
-    struct Property {
-        std::string name = "";
-        PropertyType type = PropertyType::Int;
-        std::string object_type = "";
-        std::string link_origin_property_name = "";
-        bool is_primary = false;
-        bool is_indexed = false;
-        bool is_nullable = false;
-
-        size_t table_column = -1;
-        bool requires_index() const { return is_primary || is_indexed; }
-
-        bool is_indexable() const
-        {
-            return type == PropertyType::Int
-                || type == PropertyType::Bool
-                || type == PropertyType::Date
-                || type == PropertyType::String;
-        }
+enum class PropertyType : unsigned char {
+    Int    = 0,
+    Bool   = 1,
+    String = 2,
+    Data   = 3,
+    Date   = 4,
+    Float  = 5,
+    Double = 6,
+    Object = 7,
+    LinkingObjects = 8, // Implies Array
```

Comment:
What do you mean by "Implies Array" here? Is a linking objects property always be represented as `LinkingObjects | Array`, or is `LinkingObjects` by itself always treated as an array? The implementation of `is_array` suggests the former.